File: src/Atlasd/Battlenet/Protocols/Game/Messages/SID_GETICONDATA.cs
using Atlasd.Battlenet.Exceptions;
using Atlasd.Daemon;
using System;
using System.IO;
using System.Text;

namespace Atlasd.Battlenet.Protocols.Game.Messages
{
    class SID_GETICONDATA : Message
    {
        public SID_GETICONDATA()
        {
            Id = (byte)MessageIds.SID_GETICONDATA;
            Buffer = new byte[16];
        }

        public SID_GETICONDATA(byte[] buffer)
        {
            Id = (byte)MessageIds.SID_GETICONDATA;
            Buffer = buffer;
        }

        public override bool Invoke(MessageContext context)
        {
            Logging.WriteLine(Logging.LogLevel.Debug, Logging.LogType.Client_Game, context.Client.RemoteEndPoint, $"[{Common.DirectionToString(context.Direction)}] SID_GETICONDATA ({4 + Buffer.Length} bytes)");

            switch (context.Direction)
            {
                case MessageDirection.ClientToServer:
                    {
                        if (Buffer.Length != 0)
                            throw new GameProtocolViolationException(context.Client, "SID_GETICONDATA buffer must be 0 bytes");

                        return new SID_GETICONDATA().Invoke(new MessageContext(context.Client, MessageDirection.ServerToClient));
                    }
                case MessageDirection.ServerToClient:
                    {
                        /**
                         * (FILETIME) Filetime
                         *   (STRING) Filename
                         */

                        var fileInfo = new FileInfo("icons.bni");

                        var fileTime = (UInt64)fileInfo.LastWriteTimeUtc.ToFileTimeUtc();
                        var fileName = fileInfo.Name;

                        Buffer = new byte[9 + Encoding.UTF8.GetByteCount(fileName)];

                        var m = new MemoryStream(Buffer);
                        var w = new BinaryWriter(m);

                        w.Write((UInt64)fileTime);
                        w.Write(Encoding.UTF8.GetBytes(fileName));
                        w.Write((byte)0);

                        w.Close();
                        m.Close();

                        Logging.WriteLine(Logging.LogLevel.Debug, Logging.LogType.Client_Game, context.Client.RemoteEndPoint, $"[{Common.DirectionToString(context.Direction)}] SID_GETICONDATA ({4 + Buffer.Length} bytes)");
                        context.Client.Send(ToByteArray(context.Client.ProtocolType));
                        return true;
                    }
            }

            return false;
        }
    }
}

File: src/AtlasV/Battlenet/Protocols/Game/Messages/SID_GETICONDATA.vb
Imports AtlasV.Battlenet.Exceptions
Imports AtlasV.Daemon
Imports System
Imports System.IO
Imports System.Text

Namespace AtlasV.Battlenet.Protocols.Game.Messages
    Class SID_GETICONDATA
        Inherits Message

        Public Sub New()
            Id = CByte(MessageIds.SID_GETICONDATA)
            Buffer = New Byte(15) {}
        End Sub

        Public Sub New(ByVal varBuffer As Byte())
            Id = CByte(MessageIds.SID_GETICONDATA)
            Buffer = varBuffer
        End Sub

        Public Overrides Function Invoke(ByVal context As MessageContext) As Boolean
            Logging.WriteLine(Logging.LogLevel.Debug, Logging.LogType.Client_Game, context.Client.RemoteEndPoint, $"[{Common.DirectionToString(context.Direction)}] {MessageName(Id)} ({4 + Buffer.Length} bytes)")

            Select Case context.Direction
                Case MessageDirection.ClientToServer

                    If Buffer.Length <> 0 Then
                        Throw New GameProtocolViolationException(context.Client, $"{MessageName(Id)} buffer must be 0 bytes")
                    End If

                    Return New SID_GETICONDATA().Invoke(New MessageContext(context.Client, MessageDirection.ServerToClient))
                Case MessageDirection.ServerToClient
                    Dim fileInfo = New FileInfo("icons.bni")
                    Dim fileTime = CULng(fileInfo.LastWriteTimeUtc.ToFileTimeUtc())
                    Dim fileName = fileInfo.Name
                    Buffer = New Byte(9 + Encoding.UTF8.GetByteCount(fileName) - 1) {}

                    Using m = New MemoryStream(Buffer)
                        Using w = New BinaryWriter(m)
                            w.Write(CULng(fileTime))
                            w.Write(fileName)
                        End Using
                    End Using

                    Logging.WriteLine(Logging.LogLevel.Debug, Logging.LogType.Client_Game, context.Client.RemoteEndPoint, $"[{Common.DirectionToString(context.Direction)}] {MessageName(Id)} ({4 + Buffer.Length} bytes)")
                    context.Client.Send(ToByteArray(context.Client.ProtocolType))
                    Return True
            End Select

            Return False
        End Function
    End Class
End Namespace
